### PR TITLE
Read tls client cert from local TC when TC is heterogeneous or across-kubernetes

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Read tls client cert from local TC when TC is heterogeneous or across-kubernetes

### Code changes

- [x] Has Go code change
- [ ] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 

heterogeneous:
1. Deploy a cluster with enable TLS
2. Deploy heterogeneous cluster to join

across-kubernetes:
1. Deploy a cluster with enable TLS
2.  Deploy heterogeneous cluster in other kubernetes to join

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note
Read tls cert from local TC when TC is heterogeneous or across-kubernetes
```

### CODE REVIEW

**Comment by mikechengwei** [comment]
I understand that pdclient is constructed with tls of local tc, what about tls of hetegereous cluster itself?

**Comment by KanShiori** [comment]
> I understand that pdclient is constructed with tls of local tc, what about tls of hetegereous cluster itself?

heterogeneous cluster should use itself tls client cert, as example yaml https://github.com/pingcap/tidb-operator/blob/master/examples/heterogeneous-tls/cert-manager/heterogeneous/cluster-client-secret.yaml

**Review by csuzhangxc — APPROVED** [review]

File: pkg/pdapi/pd_control.go [hunk 1]
```diff
@@ -28,20 +28,88 @@ import (
...
 	GetEndpoints(namespace Namespace, tcName string, tlsEnabled bool) (endpoints []string, tlsConfig *tls.Config, err error)
 }
 
+type clientConfig struct {
+	clusterDomain string
+
+	// clientURL is PD addr. If it is empty, will generate from target TC
+	clientURL string
+	// clientName is client name. If it is empty, will generate from target TC
+	clientName string
```
Comment (DanielZhangQD) [inline pkg/pdapi/pd_control.go:81]: This is `clientKey`, right?

File: pkg/controller/pd_control.go [hunk 2]
```diff
@@ -18,14 +18,14 @@ import (
 	"github.com/pingcap/tidb-operator/pkg/pdapi"
 )
 
-// getPDClientFromService gets the pd client from the TidbCluster
-func getPDClientFromService(pdControl pdapi.PDControlInterface, tc *v1alpha1.TidbCluster) pdapi.PDClient {
+// GetPDClientFromService gets the pd client from the TidbCluster
+func GetPDClientFromService(pdControl pdapi.PDControlInterface, tc *v1alpha1.TidbCluster) pdapi.PDClient {
 	if tc.HeterogeneousWithoutLocalPD() {
-		// TODO: to support across k8s cluster without local pd
```
Comment (DanielZhangQD) [inline pkg/controller/pd_control.go:24]: This comment should be reserved since we do not support this case yet.

File: pkg/pdapi/pd_control.go [hunk 3]
```diff
@@ -28,20 +28,88 @@ import (
...
+
+// TLSCertFromTC indicate that the client use certs from specified TC's secret.
+func TLSCertFromTC(ns Namespace, tcName string) Option {
+	return func(c *clientConfig) {
+		c.tlsSecretNamespace = ns
+		c.tlsSecretName = util.ClusterClientTLSSecretName(tcName)
+	}
+}
+
+// TLSCertFromTC indicate that client use certs from specified secret.
```
Comment (DanielZhangQD) [inline pkg/pdapi/pd_control.go:49]:
```suggestion
// TLSCertFromSecret indicates that clients use certs from specified secret.
```

File: pkg/pdapi/pd_control.go [hunk 4]
```diff
@@ -28,20 +28,88 @@ import (
 // Namespace is a newtype of a string
 type Namespace string
 
+// Option configure PDClient
```
Comment (DanielZhangQD) [inline pkg/pdapi/pd_control.go:31]:
```suggestion
// Option configures the PDClient
```

File: pkg/pdapi/pd_control.go [hunk 5]
```diff
@@ -28,20 +28,88 @@ import (
...
+type Option func(c *clientConfig)
+
+// ClusterRef set cluster domain of TC, it is used when generating PD addr from TC.
```
Comment (DanielZhangQD) [inline pkg/pdapi/pd_control.go:34]:
```suggestion
// ClusterRef sets the cluster domain of TC, it is used when generating the PD address from TC.
```

File: pkg/pdapi/pd_control.go [hunk 6]
```diff
@@ -28,20 +28,88 @@ import (
...
+func ClusterRef(clusterDomain string) Option {
+	return func(c *clientConfig) {
+		c.clusterDomain = clusterDomain
+	}
+}
+
+// TLSCertFromTC indicate that the client use certs from specified TC's secret.
```
Comment (DanielZhangQD) [inline pkg/pdapi/pd_control.go:41]:
```suggestion
// TLSCertFromTC indicates that the client uses the certs from the specified TC's secret.
```

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**